lib/code-caller-docker.js

```diff
@@ -0,0 +1,703 @@
+            ];
+            await this._createAndAttachContainer(binds);
+            this.state = WAITING;
+            this._checkState();
+            this.debug('exit ensureChild()');
+        });
+    }
+
+    async _ensureDirectoryIfNeeded() {
+        await this.ensureDirectoryMutex.runExclusive(async () => {
+            if (config.workersExecutionMode === 'container' && !this.hostDirectory) {
+                this.hostDirectory = await tmp.dir({ unsafeCleanup: true, prefix: MOUNT_DIRECTORY_PREFIX });
+            }
+        });
+    }
+
+    /**
+     * Creates a container and attaches its stdin/stdout/stderr to streams
+     * we can write to and read from.
+     * 
+     * @param {string[]} binds - The bind mounts for the container
+     */
+    async _createAndAttachContainer(binds) {
+        this.debug('enter _createAndAttachContainer');
+        this.debug('_createAndAttachContainer(): creating container');
+        this.container = await docker.createContainer({
+            name: `prairielearn.worker.${this.uuid}`,
+            Image: getExecutorImageName(),
+            OpenStdin: true,
+            // This will close stdin once we disconnect, which will let
+            // the worker know that they should die. Once the worker dies,
+            // `AutoRemove: true` below will ensure the worker container
+            // is removed.
+            StdinOnce: true,
+            AttachStdin: true,
+            AttachStdout: true,
+            AttachStderr: true,
+            HostConfig: {
+                Binds: binds,
+                AutoRemove: true,
+                // Prevent forkbombs
+                PidsLimit: 64,
+            },
+            Env: [
+                // Proxy the `DEBUG` environment variable to the container so we can
+                // turn on debug logging for it
+                `DEBUG=${process.env.DEBUG}`,
+            ],
+        });
+        this.debug('_createAndAttachContainer(): created container');
+
+        const stream = await this.container.attach({
+            stream: true,
+            stdin: true,
+            stdout: true,
+            stderr: true,
+        });
+        this.stdinStream = new MemoryStream();
+        this.stdoutStream = new MemoryStream();
+        this.stderrStream = new MemoryStream();
+        this.stdinStream.pipe(stream);
+        this.container.modem.demuxStream(stream, this.stdoutStream, this.stderrStream);
+        this.stdoutStream.on('data', data => this._handleOutput(data));
+        // TODO: we should really report stderr in logs
+        // To see stderr from the container for debugging, uncomment the following line
+        this.stderrStream.pipe(process.stderr);
+
+        this.debug('_createAndAttachContainer(): starting container');
+        await this.container.start();
+        this.debug('_createAndAttachContainer(): container started');
+
+        // This will passively wait for the container to exit or be killed
```

Comment:
Based on the resolved conversation a few lines down:
```suggestion
        // This will passively (non-blocking) wait for the container to exit or be killed, while allowing other code to keep executing
```